firmware/shared/controls/tc_scale_factor.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+namespace ctrl {
+// Note: The CalculateActualSlip function has Div-by-Zero error if left front
+// and right front wheel speeds = 0.
+template <typename T>
+T CalculateActualSlip(T left_rear_wheel_speed, T right_rear_wheel_speed,
+                      T left_front_wheel_speed, T right_front_wheel_speed) {
+    T idle_wheel_spd = (left_front_wheel_speed + right_front_wheel_speed) / 2.0;
+    T actual_slip;
+
+    if (left_rear_wheel_speed > right_rear_wheel_speed) {
+        actual_slip = (left_rear_wheel_speed / idle_wheel_spd) - 1;
+    } else {
+        actual_slip = (right_rear_wheel_speed / idle_wheel_spd) - 1;
+    }
+
+    if (actual_slip < 0) {
+        actual_slip = 0;
+    }
+    return actual_slip;
+}
+
+template <typename T>
+T CalculateTCScaleFactor(T actual_slip, T target_slip, int time_ms) {
+    // Stateflow: Multi-Stage TC
+    T scale_factor;
```

Comment:
I also had a couple more questions:
1. What is this LUT in the `simp_vd_interface` Simulink subsystem? It looks like it just scales the input by 10, am I right to assume that?
![image](https://github.com/user-attachments/assets/4dd98f62-315c-454a-a78e-d29f81ea1fca)
2. How can I test the subsystem in Simulink, so that I can generate test cases to use for the c++ implementation?

Reply:
1. Yes it looks like a multiply by ten, but that doesn't logically make sense since that implies 1000% torque. I'm guessing that it was a hack because the running average block stores 10 values.

We will implement it differently for the C++ implementation. You should use the LookupTable from `shared/util/mappers/lookup_table`. See [this file](https://github.com/macformula/racecar/blob/freerb/tms/firmware/projects/TMS/inc/tempsensor.hpp) for an example on creating a lookup table. Your table should only have two key-values: 0,0 and 100,100. This means it doesn't actually modify the input, but in the future we will want the ability to modify this curve.

Replace the entire Running Average block with MovingAverage from `shared/util/moving_average`. See [this file](https://github.com/macformula/racecar/blob/freerb/tms/firmware/projects/TMS/inc/app.hpp) for an example.

2. I'm not sure how to test simulink, but I don't really trust the model anyways. You should come up with test cases manually. For example, if driver_torque_request=0 then the output torque request should be zero. Don't worry too much about them rn. Just make some simple tests so that we have something to build on later.

Reply:
> @BlakeFreer For the `CalculateTCScaleFactor` function I assumed that the OS is handled by main.cc, and will call the `simp_vd_interface` update function at a regular interval. Time_ms will then represent time elapsed, and the TC scale factor will be returned based on that. Let me know if this is the right approach.

This is mostly correct. `time_ms` will be the absolute current time, not elapsed. If you want elapsed then make a `static` variable in your function (static variables don't get reset on successive function calls). Idk if your code actually needs the time, but it's there just in case.